omniduct/databases/exasol.py

```diff
@@ -0,0 +1,148 @@
+from __future__ import absolute_import
+
+from interface_meta import override
+
+from omniduct.utils.debug import logger
+
+from .base import DatabaseClient
+
+
+class ExasolClient(DatabaseClient):
+    """
+    This client connects to an Exasol service using the `pyexasol` python library.
+
+    Example Config
+    --------------
+
+    databases:
+        exasol_db:
+            protocol: exasol
+            host:
+              - 'localhost:8563'
+              - 'localhost:8564'
+            username: exasol_user
+            password: ****
+            schema: users
+    """
+
+    PROTOCOLS = ["exasol"]
+    DEFAULT_PORT = 8563
+    NAMESPACE_NAMES = ["schema", "table"]
+    NAMESPACE_QUOTECHAR = '"'
+    NAMESPACE_SEPARATOR = "."
+
+    @property
+    @override
+    def NAMESPACE_DEFAULT(self):
+        return {"schema": self.schema}
+
+    @override
+    def _init(self, schema=None, engine_opts=None):
+        self.__exasol = None
+
+        self.schema = schema
+        self.connection_fields += ("schema",)
+        self.engine_opts = engine_opts or {}
+
+    @override
+    def _connect(self):
+        import pyexasol
+
+        logger.info("Connecting to Exasol ...")
+        self.__exasol = pyexasol.connect(
+            dsn="{host}:{port}".format(host=self.host, port=self.port),
+            user=self.username,
+            password=self.password,
+            **self.engine_opts
+        )
+
+    @override
+    def _is_connected(self):
+        return self.__exasol is not None
+
+    @override
+    def _disconnect(self):
+        try:
+            self.__exasol.close()
+        except Exception:
+            pass
+        self.__exasol = None
+        self.schema = None
+
+    @override
+    def _execute(self, statement, cursor, wait, session_properties, query=True):
+        #: pyexasol.ExaStatement has a similar interface to that of
+        # a DBAPI2 cursor.
+        cursor = cursor or self.__exasol.execute(statement)
+
+        # hacky: make the result look like a cursor.
+        # cursor.columns returns a dict we transform it into a
+        # flat list.
+        cursor.description = []
+        for key, value in cursor.columns().items():
+            description = [key]
+            description.extend(value.values())
```

Comment:
The order of these descriptors matters. Have you looked into what the description looks like after doing this? If the descriptors returned by `.columns()` are not 100% overlapping with that of a DBAPI2 schema, you should manually construct them, rather than just appending values. You can read more about the fields and what they mean here: https://www.python.org/dev/peps/pep-0249/#description

Reply:
Only the first two descriptors are mandatory so I will investigate to ensure that these overlap exactly and that the order is deterministic.

I need to investigate what is expected in `type_code` since I am not sure if this need to be an `int` as with the case with postgres. As far as I am aware Exasol doesn't have an `int` to indicate their types. But I will also follow up on this point.